### ISSUE TITLE
Lib crashes while parsing info dictionary

### ISSUE THREAD

**Issue by barata04** [galkahana/PDF-Writer#136]
While trying to get all the information from the info dictionary of a certain pdf, the lib crashes. As far as i can understand, there are some properties on the info dictionary that are not correctly built (even an online pdf validator states that the pdf is not using the standards). However, the outcome should be different than a crash. Can you look into this? I've attached the pdf. The problem is when dictionary parsing goes wrong, the NotifyIndirectObjectEnd function is still called with a NULL object, which then crashes on the GetType function.
[Computer_Systems-A_Programmers_Perspective-2e.pdf](https://github.com/galkahana/PDF-Writer/files/2553679/Computer_Systems-A_Programmers_Perspective-2e.pdf)

**Comment by galkahana** [comment]
this whole part is a mess:
```
21242 0 obj<</Universal PDF(The process that creates this PDF constitutes a trade secret of codeMantra, LLC and is protected by the copyright laws of the United States)/Producer(Acrobat Distiller 7.0.5 for Macintosh)/Creator(Acrobat: pictwpstops filter 1.0)/Author(David R. O'Hallaron Randal E. Bryant)/codeMantra, LLC(http://www.codemantra.com)/ModDate(D:20110114154254-05'00')/Title(Computer Systems: A Programmer's Perspective)/CreationDate(D:20100201173206+05'30')/EBX_PUBLISHER/Prentice#20Hall>>endobj
```

that "PDF" after universal kinda makes this impossible to read. there's no parsing of this object without failing.

i'll fix the crash, but reading the dict is impossibly

**Comment by galkahana** [comment]
actually, should be checking for nulls. that's the errors system. where in the code/samples is gettype called without first checking that it got something?

**Comment by barata04** [comment]
Yeah, I understand that the dictionary is malformed. However, it should not
crash applications. As far as I can remember, it it on the
"ParseExistingIndirectObject" or something like that, at the end of the
function, where it uses the object parsed. However, as the object was not
correctly parsed, the function should not continue and return an error.

Thanks for the help!

<http://www.avg.com/email-signature?utm_medium=email&utm_source=link&utm_campaign=sig-email&utm_content=webmail>
Sem
vírus. www.avg.com
<http://www.avg.com/email-signature?utm_medium=email&utm_source=link&utm_campaign=sig-email&utm_content=webmail>
<#DAB4FAD8-2DD7-40BB-A1B8-4E2AA1F9FDF2>

gal kahana <notifications@github.com> escreveu no dia sábado, 8/12/2018
à(s) 12:19:

**Comment by galkahana** [comment]
well, reason im asking is cause i cant find where it does use the object.
im trying to parse the file and am succeeding

**Comment by galkahana** [comment]
using the JS counterpart to quickly run this, here's what im doing:
```
const hummus = require('../hummus')


var pdfReader = hummus.createReader('./Computer_Systems-A_Programmers_Perspective-2e.pdf');
var info = pdfReader.queryDictionaryObject(pdfReader.getTrailer(), "Info");

console.log('info', info)
console.log('pdf pages count', pdfReader.getPagesCount())
```

so i get null for the info, due to unsuccessful parsing,
but i do get the correct pages count later.

**Comment by barata04** [comment]
On the C++ library, the problem is at PDFParser.cpp, at the function "PDFObject* PDFParser::ParseExistingInDirectObject(ObjectIDType inObjectID)".

At the end of the method, we can find this:

"NotifyIndirectObjectStart(inObjectID, versionObject->GetValue());
readObject = mObjectParser.ParseNewObject();

NotifyIndirectObjectEnd(readObject);"

On that specific file, "readObject" is NULL at this moment and the "NotifyIndirectObjectEnd" call will crash the application.

**Comment by galkahana** [comment]
NotifyIndirectObjectEnd doesn't crash. i checked. do you have an extender that requires checking for null maybe?
